### ISSUE TITLE
initrd.secrets of previous generations prevent to switch to new generation

### ISSUE THREAD

**Issue by symphorien** [NixOS/nixpkgs#114594]
## Issue description

I have a FDE with luks. Grub prompts me for my passphrase, reads /boot, and then linux boots. To type my passphrase only once, my luks partition can also be opened by a keyfile:
```
initrd.secrets = {
      "/key" = "/boot/custom/key";
    };
    initrd.luks.devices = {
      cleartext = {
        device="/dev/disk/by-uuid/fc78a593-1003-4542-94b5-a12bd668720d";
        keyFile="/key";
        fallbackToPassword=true;
      };
```
This keyfile is optional.

I accidentally wiped /boot and had to recreate a keyfile. For some reason I put the new keyfile at a different path, but then I could not switch to the new generation, because grub could only be installed if the keyfile of old generations was available as well. I was forced to copy the keyfile at the old location as well to switch. In this case the two files are interchangeable, to it does not harm, but in some other situation, if a secret is irrevocably lost, there should be a way to ignore the missing secret, even if this makes generations unbootable.

### Steps to reproduce
Create `/boot/foo`
Switch to a generation with `initrd.secrets."/filename" = "/boot/foo"`
Move `/boot/foo` to `/boot/bar`
Switch to a generation with `initrd.secrets."/filename" = "/boot/bar"`

Switching fails because it requires both `/boot/foo` and `/boot/bar`.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
 - system: `"x86_64-linux"`
 - host os: `Linux 5.9.16, NixOS, 20.09.2750.85abeab48b5 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-20.09.2750.85abeab48b5, nixos-hardware, nixos-unstable-21.03pre265961.891f607d530"`
 - channels(symphorien): `"home-manager-20.09"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Related to https://github.com/NixOS/nixpkgs/issues/101462

**Comment by Artturin** [comment]
garbage collecting the previous generations is a workaround

**Comment by symphorien** [comment]
Well in my case I couldn't beause it also affected the generation I booted in, and one cannot remote the currently booted generation.

**Comment by SuperSandro2000** [comment]
I ran into the same issue today while installing my nixos system and having temporarily ``boot.initrd.network.ssh.*`` set. I am not sure why the script is not properly regenerated and impure.

**Comment by Sepero** [comment]
This seems to fix the issue for me, allowing nixos-rebuild. Though there appears to still be some other cruft left over.

    sudo mount -o remount,rw /nix/store
    sudo find /nix/store -type l -name "append-initrd-secrets" -exec mv {} {}.bak \;

Source: #101462

**Comment by Sepero** [comment]
Follow-up: It appears that I am able to clear out the cruft by deleting previous generations.

**Comment by pwaller** [comment]
I think there is a bug here.

TLDR; The secrets are only named after the initrd, so if the initrd is unchanged, as install-grub.pl runs, it overwrites the secrets in turn with the secrets in the system corresponding oldest matching initrd name.

Profiles are added in a loop here:

https://github.com/NixOS/nixpkgs/blob/d47adb096e27c6d843d8242ad967114768aa64c4/nixos/modules/system/boot/loader/grub/install-grub.pl#L571

This appears to add generations in the most-recent-first order.

Then `addEntry` is responsible for creating a secrets for the profile, which it does by renaming onto `$initrdSecretsPath`.

https://github.com/NixOS/nixpkgs/blob/d47adb096e27c6d843d8242ad967114768aa64c4/nixos/modules/system/boot/loader/grub/install-grub.pl#L464

However, this path is only a function of the initrd name:

https://github.com/NixOS/nixpkgs/blob/d47adb096e27c6d843d8242ad967114768aa64c4/nixos/modules/system/boot/loader/grub/install-grub.pl#L454

Which is the base name of the `Cwd::abs_path`:

https://github.com/NixOS/nixpkgs/blob/d47adb096e27c6d843d8242ad967114768aa64c4/nixos/modules/system/boot/loader/grub/install-grub.pl#L449

Consequently, if the initrd is unchanged between generations, it looks to me as though the secrets will end up using the oldest version of the secrets available for that initrd.

Indeed if I look through my grub menu, that's what I see. When I change the secrets, I get a new `append-initrd-secrets`, but not a new `/boot/kernels/*-initrd-secrets`.

**Comment by pwaller** [comment]
I've posted a potential fix for the above described issue at https://github.com/NixOS/nixpkgs/pull/209156.

**Comment by symphorien** [comment]
https://github.com/NixOS/nixpkgs/pull/209156 does not fully fix this issue: renaming a secret prevents running nixos-rebuild as long as there are still generations that reference it. However if the secret is lost then there is no way we can reconstruct a sensible initird, so I think the workaround of `touch`ing the lost file is good enough.
